File: go/graph.go
package main

type Graph struct {
    adjList [][]int // Adjacency list representation of the graph
}

func NewGraph(numVertices int) *Graph {
    adjList := make([][]int, numVertices)
    return &Graph{adjList}
}

func (g *Graph) AddEdge(u, v int) {
	// Add an undirected edge between nodes u and v
    g.adjList[u] = append(g.adjList[u], v)
    g.adjList[v] = append(g.adjList[v], u)
}

func (g *Graph) BFS(start int) []int {
    visited := make([]bool, len(g.adjList)) // Keep track of visited nodes
    queue := make([]int, 0) // FIFO queue for BFS traversal
    traversal := make([]int, 0) // Store the traversal order

    visited[start] = true
    queue = append(queue, start)

    for len(queue) > 0 {
        node := queue[0]
        queue = queue[1:]
        traversal = append(traversal, node)

		// Visit each neighbor of the current node that has not been visited yet
        for _, neighbor := range g.adjList[node] {
            if !visited[neighbor] {
                visited[neighbor] = true
                queue = append(queue, neighbor)
            }
        }
    }

    return traversal
}

func exampleGraph() *Graph {
	graph := NewGraph(6) // Create a new graph with 6 nodes
    graph.AddEdge(0, 1) // Add edges between the nodes
    graph.AddEdge(0, 2)
    graph.AddEdge(1, 3)
    graph.AddEdge(2, 3)
    graph.AddEdge(2, 4)
    graph.AddEdge(3, 4)
    graph.AddEdge(3, 5)

	return graph
}

File: rust/src/graph.rs
use std::collections::VecDeque;

pub struct Graph {
    adj_list: Vec<Vec<usize>>, // Adjacency list representation of the graph
}

impl Graph {
    fn new(num_vertices: usize) -> Self {
        Graph {
            adj_list: vec![Vec::new(); num_vertices],
        }
    }

    fn add_edge(&mut self, u: usize, v: usize) {
        // Add an undirected edge between nodes u and v
        self.adj_list[u].push(v);
        self.adj_list[v].push(u);
    }

    pub fn bfs(&self, start: usize) -> Vec<usize> {
        let mut visited = vec![false; self.adj_list.len()]; // Keep track of visited nodes
        let mut queue = VecDeque::new(); // FIFO queue for BFS traversal
        let mut traversal = Vec::new(); // Store the traversal order

        visited[start] = true;
        queue.push_back(start);

        while let Some(node) = queue.pop_front() {
            traversal.push(node);
            // Visit each neighbor of the current node that has not been visited yet
            for neighbor in &self.adj_list[node] {
                if !visited[*neighbor] {
                    visited[*neighbor] = true;
                    queue.push_back(*neighbor);
                }
            }
        }

        traversal
    }
}

pub fn example_graph() -> Graph {
    let mut graph = Graph::new(6);
    graph.add_edge(0, 1);
    graph.add_edge(0, 2);
    graph.add_edge(1, 3);
    graph.add_edge(2, 3);
    graph.add_edge(2, 4);
    graph.add_edge(3, 4);
    graph.add_edge(3, 5);
    return graph;
}